### ISSUE TITLE
Unable to load big data files

### ISSUE THREAD

**Issue by sargdavid** [hadley/haven#92]
Hello Hadley,

I'm having difficulties loading a 3 GB .sas7dat file using haven::read_sas. The file opens in SAS but I get the following error in R:

   Error: Failed to parse C:\XX\xx.sas7bdat: Invalid file, or file has unsupported features.

The table has 15 million + rows and 48 columns.

As a work-around, I used your other package - readr. I exported my SAS file as a .CSV, then used read_csv to read it into R. It would be better, of course, if we could import data directly with read_sas. Is there a way to make it work or is it a limitation of this function?

NOTE: read_csv imported the table correctly but with the following warning/problem:

   Warning message:
   8974944 problems parsing 'xx.csv'. See problems(...) for more details.

   Error in withCallingHandlers(expr, warning = function(w) invokeRestart("muffleWarning")) : 
     argument "x" is missing, with no default

I don't know if that would help...

Thank you!

**Comment by hadley** [comment]
It's probably not an issue with the file size, but some feature that the file uses. Can you reproduce the error when saving a smaller subset of the data that you can share?

**Comment by sargdavid** [comment]
Thank you, let me try that and get back to you...

Regards,

Davit

**Comment by adulbrich** [comment]
Dear Hadley,

Following the post from sargdavid, I had similar issues while trying to import large sas7bdat datasets (> 3 GB approximately). I tried several things:

- importing a 2 GB file: works
- importing 4 and 12 GB files: does not work
- importing a small sample of the 12 GB file: works

Here is the Error while running the code:

```R
> system.time(test.big <- read_sas(fileNameOcv))
Error: Failed to parse D:\Data\hh.sas7bdat: Unable to read file.
Timing stopped at: 0 0 0.06
```

And here is the code I used for benchmarking purposes (I love your package by the way, it's even faster than RevoScaleR):

```R
library(haven)
library(RevoScaleR)

############################################################################
##### 1.81 GB (4050025 rows, 57 columns)
fileName <- "D:/Data/bsn.sas7bdat"

### File Import

system.time(test.small <- read_sas(fileName))
# {haven}
# user  system elapsed 
# 283.2     1.5   284.8 
# user  system elapsed 
# 345.98    1.48  347.64

system.time(test.small <- rxImport(fileName))
# {RevoScaleR}
# user  system elapsed 
# 37.78    1.84 1122.17
# user  system elapsed 
# 51.28    1.47 1198.97

############################################################################
##### 11.9GB (2134616 rows, 652 columns)
fileNameOcv <- "D:/Data/hh.sas7bdat"

system.time(test.big <- read_sas(fileNameOcv))

system.time(test.big <- rxImport(fileNameOcv))

```

FYI, the rxImport function is able to read the file, but takes an incredible amount of time to process it once read, as it consumes a lot of RAM.

Thank you for your help,

Kind regards,

Alexander

**Comment by hadley** [comment]
@adulbrich unfortunately it's impossible to help without the file itself or otherwise a reproducible example. The error message suggests that it's unlikely to be the size of the file alone.

**Comment by adulbrich** [comment]
@hadley as for the reproducible example, I've generated a huge matrix filed mostly with zeros (but for the first column and line), which is approximately 7 GB in size (2M rows, 600 columns). It gives the same error as previously stated when I try to import the 7GB file, but it works when I take a sample consisting of the first 1000 observations. 

The matrix was generated in R, and imported in SAS BASE 9.4 to create a sas7bdat file. Both files (the 7 GB and its sample) are uncompressed, and have the same attributes.

**Comment by hadley** [comment]
Thanks! That suggests it is an indeed a ReadStat bug. Could you try creating the smallest possible file that illustrates the problem? @evanmiller will need to take a look.

**Comment by adulbrich** [comment]
@evanmiller @hadley I've managed to shrink down the file down to 2.4 GB and the issue still appears, but the message is different this time. Those are all samples from the same 7 GB file.
```R
> system.time(test.big <- read_sas(fileNameTestMatSamp)) # 2.4 GB
Error: Failed to parse D:\Data\testmatrix_sample.sas7bdat: Invalid file, or file has unsupported features.
Timing stopped at: 0.02 0.01 0.06 
>
> system.time(test.big <- read_sas(fileNameTestMat)) # 7 GB
Error: Failed to parse D:\Data\testmatrix.sas7bdat: Unable to read file.
Timing stopped at: 0 0.02 0.02 
>
> system.time(test.big <- read_sas(fileNameTestMatSamp)) # 2 GB
   user  system elapsed 
 293.98    1.75  296.59 
```
The 2 GB file does work however. I'll take the generated 2.4 GB data home with me this weekend. That way it will be easier to transfer it to you guys.

**Comment by adulbrich** [comment]
Sorry for the delay;

Sample that does work:
https://mega.nz/#!QYJgUYpK!TDlduNBdEPLg3rId3XEgX2Zi1A0sFb4VC6ZNrKbDmX0

Sample that does not work:
https://mega.nz/#!RFJGyZhZ!AuoZby304Xsr66UsISKG0MWCpvRR73F-PZd9wVjdk3I

**Comment by evanmiller** [comment]
2nd file works fine for me with ReadStat, will test 1st file.

**Comment by adulbrich** [comment]
I've tested the package on 3 different computers running Win7, Win8.1, WinServer2012R2, through an AFS repository and on a local hard drive, unsuccessfully. I am not using the GitHub development version though. I'll try that at home.

**Comment by hadley** [comment]
@adulbrich you don't need to do anything else. I'll need to take a look when I'm next working on haven

**Comment by evanmiller** [comment]
Update: I recently fixed a similar issue in ReadStat (https://github.com/WizardMac/ReadStat/issues/34). Here are instructions for using the development version of the ReadStat code:

https://github.com/hadley/haven#updating-readstat

The issue was preventing large files from being opened on Windows (and AIX). The latest code has been reported to work on files as large as 9.7 GB on Windows, so try it out and let me know if it works for you.

**Comment by evanmiller** [comment]
Should be fixed by #114.
